### DISCUSSION TITLE
Is there any possible way to change /api directory?

### DISCUSSION THREAD

**Discussion by ilsubyeega** [vercel/vercel#6598] [Help] [answered]
When developing serverless using Vercel, api directory is used in root directory. I still haven't found a way to change this directory(a API directory). Can I get an approach if changing API directory is possible?

Example) Changing `/api` directory to `/server/api` etc.
Thanks.

**Comment by leerob** [comment, marked as answer]
It is not possible to change the `/api` directory. However, you can [rewrite](https://vercel.com/docs/configuration#project/rewrites) the requests!

**vercel.json**
```json
{
  "rewrites": [
    { "source": "/server/api", "destination": "/api" }
  ]
}
```

If you're using Next.js, this would be inside [next.config.js](https://nextjs.org/docs/api-reference/next.config.js/rewrites) instead.

**Reply by ilsubyeega** [reply]
Well, thats rip, thanks for answering!

**Reply by lolusername** [reply]
@leerob Hello,

I am using NUXT and I have my  project in a repo with other parts of the project (a sanity cms backend). So I can't have API be on the root, is there anyway to deploy the APIs if the API folder is in `/web/api` instead of `/api`

**Reply by ryanmstokes** [reply]
I second this question. I want to have my projects api functions in a different directory. How can we rewrite it the other way round?

 { "source": "/api", "destination": "/server/api" }

??

**Reply by jameslawson** [reply]
I'm also looking for this feature as we have a large monorepo with many strict directory conventions to allow it to scale easily. 

While I understand vercel is going for "convention over config" and promoting zero-config deploys, the funny thing is that we'd actually end up writing *more* config to get vercel functions put into my repo, because the location of the folder is just not within *our* repo's conventions. 

Netflify allow us to put our functions where we want: https://docs.netlify.com/functions/optional-configuration/?fn-language=ts#directory and is one of the reasons I chose Netlify over Vercel in the end. This comment is not for me to bash Vercel, but rather to kindly ask Vercel to consider this config to their roadmap. Netlify put it as one of their "top pieces of config" alongside build command, build directory, root directory - and with functions being very core piece to many jamstack deploys I think adding an option to vercel.json would help many monorepo maintainers. Thank you!

**Reply by leerob** [reply]
> and with functions being very core piece to many jamstack deploys I think adding an option to vercel.json would help many monorepo maintainers

It's the same amount of configuration to `rewrite` from `/api` to `/my-other-path` as it is to define a separate folder for your functions 🙏

**Reply by eraoul** [reply]
Thinking of switching from Render to Vercel, but this is probably a deal-breaker. Already have a large monorepo with Prisma-related API stuff in /api and a separate python backend for other functions in a different directory. Too much refactoring to make the JS and python libraries switch places, I'm afraid.

**Comment by zsiegel92** [comment]
I believe it is possible, and in the below case it was necessary, to use a directory other than `/api/` for serverless functions:

- NextJS (which Vercel owns) currently uses the `/app/` folder for file-based routing.
- The most popular NextJS auth framework, Next-Auth, "exposes a REST API" on routes such as `/api/auth/register`
- This implies that routes such as `/api/auth/...` should be served from `/app/api/auth/...`
- The default Vercel configuration for discoverable serverless endpoints is a file in the `/api/` directory. The default configuration serves these endpoints with the prefix `/api/`.
- There is ambiguity - should `/api/` routes be served from handlers in the `/api/` directory or in `/app/api/` by NextJS?

The way I've found to resolve this conflict:

- Use the "Legacy"/deprecated [`builds`](https://vercel.com/docs/projects/project-configuration#builds) and [`routes`](https://vercel.com/docs/projects/project-configuration#routes) parameters in `vercel.json`: Say I have a Python function in `/py/index.py` (can't go in `/api/` without a rewrite to avoid conflict with NextJS `/app/api` routes), I can do the equivalent of "changing the `/api` directory" with the following:
```json
  "builds": [
    {
      "src": "py/index.py",
      "use": "@vercel/python"
    }
  ],
  "routes": [{ "src": "/py/(.*)", "dest": "py/index.py" }]
```

I think this can also be solved with the [`functions`](https://vercel.com/docs/projects/project-configuration#functions) parameters in `vercel.json`.
